From: Eduardo Habkost <ehabkost@redhat.com>
Date: Fri, 12 Feb 2010 15:00:10 -0500
Subject: [kvm] PIT: control word is write-only
Message-id: <20100212150010.GO22381@blackpad.lan.raisama.net>
Patchwork-id: 23246
O-Subject: [RHEL6 PATCH] KVM: CVE-2010-0309: PIT: control word is write-only
Bugzilla: 560905
CVE: CVE-2010-0309
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Juan Quintela <quintela@redhat.com>
RH-Acked-by: Avi Kivity <avi@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=560905
CVE: CVE-2010-0309

The patch already entered the RHEL 5.5 and 5.4.z KVM branches.

Cherry-pick of upstream commit:

commit ee73f656a604d5aa9df86a97102e4e462dd79924
Author: Marcelo Tosatti <mtosatti@redhat.com>
Date:   Fri Jan 29 17:28:41 2010 -0200

    KVM: PIT: control word is write-only

    PIT control word (address 0x43) is write-only, reads are undefined.

    Cc: stable@kernel.org
    Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>

Signed-off-by: Eduardo Habkost <ehabkost@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/arch/x86/kvm/i8254.c b/arch/x86/kvm/i8254.c
index 23874fc..3c4f8f6 100644
--- a/arch/x86/kvm/i8254.c
+++ b/arch/x86/kvm/i8254.c
@@ -465,6 +465,9 @@ static int pit_ioport_read(struct kvm_io_device *this,
 		return -EOPNOTSUPP;
 
 	addr &= KVM_PIT_CHANNEL_MASK;
+	if (addr == 3)
+		return 0;
+
 	s = &pit_state->channels[addr];
 
 	mutex_lock(&pit_state->lock);
